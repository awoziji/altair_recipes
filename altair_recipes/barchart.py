"""Generate barcharts."""
from .signatures import bivariate_recipe, color
import altair as alt
from autosig import autosig, Signature


@autosig(bivariate_recipe + Signature(color=color(default=None, position=3)))
def barchart(data, x=0, y=1, color=None, height=300, width=400):
    """Generate a barchart."""
    chart = alt.Chart(
        data,
        height=height,
        width=width / ((len(data[x].unique()) if color is not None else 1)),
    ).mark_bar()
    if color is None:
        return chart.encode(x=x, y=y)
    else:
        return chart.encode(x=alt.X(color, axis=None), y=y, color=color, column=x)
