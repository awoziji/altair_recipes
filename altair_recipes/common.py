import altair as alt
from autosig import signature, Signature, param
import pandas as pd


@signature
class Recipe(Signature):
    data = param()
    mark = param(default={}, convert=dict)
    encoding = param(default={}, convert=dict)
    properties = param(default={}, convert=dict)


@signature
class UnivariateRecipe(Recipe):
    column = param(default=0)


@signature
class BivariateRecipe(Recipe):
    x = param(default="x")
    y = param(default="y")


@signature
class MultivariateRecipe(Recipe):
    columns = param(default=None)
    group_by = param(default=None)

    def validate(self):
        if self.columns is None: columns = list(self.data.columns)


def viz_reg_test(test_f):
    def fun(regtest):
        plot = test_f()
        if regtest is not None:
            regtest.write(plot.to_json())
        plot.save(
            test_f.__code__.co_filename + "_" + test_f.__qualname__ + ".html")
        return plot

    return fun


def to_dataframe(data):
    #this is an undocumented hack
    return alt.Chart(data).data


def default(*args):
    return next(x for x in args if x is not None)


# unused should kill?
# def update(d1, d2):
#     d1.update(d2)
#     return d1


def gather(data, key, value, columns):
    return pd.melt(
        data,
        id_vars=[col for col in data.columns if col not in columns],
        value_vars=columns,
        var_name=key,
        value_name=value)
