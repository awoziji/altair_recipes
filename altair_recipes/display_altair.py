from IPython.core.display import display, HTML


def display_html(x):
    return display(HTML(x))


display_html("""
<script src="https://cdn.jsdelivr.net/npm/vega@3"></script>
<script src="https://cdn.jsdelivr.net/npm/vega-lite@2"></script>
<script src="https://cdn.jsdelivr.net/npm/vega-embed@3"></script>
""")


def display_altair_chart(x):
    display_html("""
 <div id="vis"></div>
  <script type="text/javascript">
    var spec = """ + x.to_json() + """;
    var opt = {"renderer": "canvas", "actions": false};
    vegaEmbed("#vis", spec, opt);
  </script>
  """)
