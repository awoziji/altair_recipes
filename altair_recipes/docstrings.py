"""Reusable docstrings elements."""
from enum import Enum


class Docstring(Enum):
    """Reusable docstring elements collected in an enum."""

    data = """data: Altair Data or pandas DataFrame or csv or json file URL
    The data from which the statistical graphics is being generated"""
    column = """column: str or other column selector
    The column containing the data to be used in the graphics"""
    x = """x: str or other columns selector
    The column containing the data associated with the horizontal dimension"""
    y = """y: str or other column selector
    The column containing the data associated with the vertical dimension"""
    columns = """columns: str or list thereof or other column selector
    The column or columns to be used in the graphics"""
    group_by = """group_by: str or other column selector
    The column to be used to group the data when in long form. When group_by is
    specified columns should point to a single column"""
    return_object = """altair.Chart_area
    The graphics object"""
    mark = """mark: dictionary
    Additional arguments to pass to the mark method of Chart"""
    encoding = """encoding: dictionary
    Additional arguments to the encode method of Chart"""
    properties = """properties: dictionary
    Additional arguments to the properties method of Chart"""
    returns = """altair.Chart
    The chart described in the summary"""


def make_docstring(summary, params, returns=Docstring.returns):
    """Make docstrings from simple reusable parts.

    Parameters
    ----------
    summary : str
        The docstring summary.
    params : str
        A list of parameter docstrings. If present in the .docstrings.Docstring
        enum, will be replaced by corresponding value (default meaning for that
        argument)
    returns : str
        The docstring return section. If just "returns", will be replaced by
        .docstrings.Docstring.returns

    Returns
    -------
    str
        A docstring. Assign to __doc__ of desired object.

    """
    params = list(
        map(lambda x: x.value if type(x) == Docstring else x, params))
    returns = returns.value if type(returns) == Docstring else returns
    return "    \n".join([summary, "\nParameters\n---------"] + params +
                         ["\nReturns\n-------", returns])
