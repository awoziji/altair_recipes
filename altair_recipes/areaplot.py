"""areaplots."""
from common import warn_not_distinct
from .signatures import bivariate_recipe, color
import altair as alt
from autosig import autosig, Signature, param
from logging import warning
from enum import Enum


StackType = Enum("StackType", "true false  normalize")


@autosig(
    bivariate_recipe
    + Signature(
        color(default=2, position=3),
        stack=param(
            default=None,
            position=4,
            converter=lambda x: StackType(x) if x is not None else x,
        ),
    )
)
def areaplot(data=None, x=0, y=1, color=2, stack=None, height=600, width=800):
    """Generate a areaplot."""
    warn_not_distinct(data[x])
    if len(set(data[x])) != len(data[x]):
        warning("The relation to plot is not a function")
    if stack is not None:
        x = alt.X(x=x, stack=stack.name)
    return alt.Chart(data=data, height=height, width=width).mark_area().encode(x=x, y=y)
