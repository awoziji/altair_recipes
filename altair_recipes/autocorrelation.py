from .common import to_dataframe, UnivariateRecipe
import altair as alt
from autosig import autosig, signature, param
import numpy as np
import pandas as pd


@signature
class Autocorrelation(UnivariateRecipe):
    max_lag = param(default=None)


@autosig(Autocorrelation)
def autocorrelation(data,
                    column=0,
                    max_lag=None,
                    mark={},
                    encoding={},
                    properties={}):
    data = to_dataframe(data)
    max_lag = data.shape[0] - 1 if max_lag is None else int(max_lag)
    if type(column) is int:
        column = data.columns[0]
    lags = np.arange(0, max_lag + 1)
    _data = pd.DataFrame(
        dict(
            Lag=lags,
            Autocorrelation=[data[column].autocorr(lag=lag) for lag in lags]))
    return alt.Chart(_data).mark_bar(**mark).encode(
        x="Lag:O", y="Autocorrelation" + ":Q",
        **encoding).properties(**properties)
