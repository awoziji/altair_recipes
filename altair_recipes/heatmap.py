"""Heatmap implementation."""
import altair as alt
from .docstrings import make_docstring
from .common import update_kwargs as uk, to_dataframe
from math import sqrt


def heatmap(data, x, y, color, mark={}, encoding={}, properties={}):
    """See below."""
    data = to_dataframe(data)
    nrow = data.shape[0]
    return alt.Chart(data).mark_rect(**mark).encode(**uk(
        x=alt.X(x + ':Q', bin=alt.Bin(maxbins=3 * sqrt(nrow / 18))),
        y=alt.Y(y + ':Q', bin=alt.Bin(maxbins=2 * sqrt(nrow / 18))),
        color=alt.Color(
            'average(' + color + '):Q', scale=alt.Scale(scheme='greenblue')),
        updates=encoding)).properties(**properties)


heatmap.__doc__ = make_docstring(
    heatmap,
    """Generate a heatmap.""",
    additional_params=dict(color="""color: `str` or other column selector
    The color to be used at each `(x,y)` location"""))


def binned_heatmap(data, x, y, mark={}, encoding={}, properties={}):
    """See below."""
    data = to_dataframe(data)
    nrow = data.shape[0]
    return alt.Chart(data).mark_rect(**mark).encode(**uk(
        x=alt.X(x + ':Q', bin=alt.Bin(maxbins=3 * sqrt(nrow / 18))),
        y=alt.Y(y + ':Q', bin=alt.Bin(maxbins=2 * sqrt(nrow / 18))),
        color=alt.Color(
            'count(' + x + '):Q', scale=alt.Scale(scheme='greenblue')),
        updates=encoding)).properties(**properties)


heatmap.__doc__ = make_docstring(
    binned_heatmap,
    summary="Create a heatmap of the count of points in each square")
