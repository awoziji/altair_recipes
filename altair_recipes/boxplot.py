"""Boxplot implementation"""
import altair as alt
from altair_recipes.common import default, gather, to_dataframe


#TODO: inject addional option into graph
#TODO: unclear how to do it with complex graphics such as this, if there is a way
def boxplot(data, columns, group_by=None, mark={}, encoding={}, properties={}):
    data = to_dataframe(data)
    assert type(columns) == str or len(columns) == 1 or group_by is None
    # we are doing wide or long format but not both
    if group_by is None:  #convert wide to long
        x = default(data.columns.name, "variable")
        y = "value"
        data = gather(data, key=x, value=y, columns=columns)
    else:
        x = group_by
        y = columns if type(columns) is str else columns[0]
    #long form assumed from here
    chart = alt.Chart(data)
    chart_bar = chart.mark_bar(filled=False)
    chart_tick = chart.mark_tick()
    min_y = "min(" + y + ")"
    max_y = "max(" + y + ")"
    median_y = "median(" + y + ")"
    q1_bar = chart_bar.encode(x=x, y="q1(" + y + ")", y2=median_y)
    q3_bar = chart_bar.encode(x=x, y=median_y, y2="q3(" + y + ")")

    min_tick = chart_tick.encode(x=x, y=min_y)
    max_tick = chart_tick.encode(x=x, y=max_y)
    rule = chart.mark_rule().encode(
        x=x, y=alt.Y(min_y, axis=alt.Axis(title=y)), y2=max_y)
    return q1_bar + q3_bar + min_tick + max_tick + rule
