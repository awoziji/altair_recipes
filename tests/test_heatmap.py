import altair_recipes as ar
from altair_recipes.common import viz_reg_test
import numpy as np
import pandas as pd

# Compute x^2 + y^2 across a 2D grid
x, y = np.meshgrid(range(-5, 5), range(-5, 5))
z = x**2 + y**2

# Convert this grid to columnar data expected by Altair
data = pd.DataFrame({'x': x.ravel(), 'y': y.ravel(), 'z': z.ravel()})


@viz_reg_test
def test_heatmap():
    return ar.heatmap(data, x='x', y='y', color='z')
