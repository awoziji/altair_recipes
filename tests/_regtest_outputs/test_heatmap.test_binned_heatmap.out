{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.4.3.json",
  "config": {
    "view": {
      "height": 300,
      "width": 400
    }
  },
  "data": {
    "url": "https://vega.github.io/vega-datasets/data/movies.json"
  },
  "encoding": {
    "color": {
      "aggregate": "count",
      "field": "IMDB_Rating",
      "type": "quantitative"
    },
    "x": {
      "bin": {},
      "field": "IMDB_Rating",
      "type": "quantitative"
    },
    "y": {
      "bin": {},
      "field": "Rotten_Tomatoes_Rating",
      "type": "quantitative"
    }
  },
  "mark": "rect"
}